File: src/lib/rate-limit.ts
import { LRUCache } from 'lru-cache'

interface Bucket {
  count: number
  resetAt: number
}

const buckets = new LRUCache<string, Bucket>({
  max: 10000,
  ttl: 1000 * 60 * 60, // safety cap; real expiry is per-window below
})

export interface RateLimitResult {
  allowed: boolean
  remaining: number
  resetAt: number
}

export function rateLimit(
  key: string,
  limit: number,
  windowMs: number,
  now: number = Date.now()
): RateLimitResult {
  const existing = buckets.get(key)

  if (!existing || existing.resetAt <= now) {
    const resetAt = now + windowMs
    buckets.set(key, { count: 1, resetAt }, { ttl: windowMs })
    return { allowed: true, remaining: limit - 1, resetAt }
  }

  if (existing.count >= limit) {
    return { allowed: false, remaining: 0, resetAt: existing.resetAt }
  }

  existing.count += 1
  return {
    allowed: true,
    remaining: limit - existing.count,
    resetAt: existing.resetAt,
  }
}

export function getClientIp(request: Request): string {
  const xff = request.headers.get('x-forwarded-for')
  if (xff) return xff.split(',')[0].trim()
  const real = request.headers.get('x-real-ip')
  if (real) return real.trim()
  return 'unknown'
}

// Test helper: clear all buckets between tests.
export function resetRateLimit(): void {
  buckets.clear()
}

File: djclass_overlay/common/ratelimit.py
"""In-memory per-IP fixed-window rate limiter. Single-process (matches
uvicorn --workers 1, like common/cache.py). Behind the Cloudflare Tunnel the
real client IP arrives in CF-Connecting-IP, not REMOTE_ADDR.
"""

import time

_MAX_KEYS = 10000
_buckets = {}  # (scope, ip) -> [window_start, count, window]


def reset():
    """Clear all buckets (tests)."""
    _buckets.clear()


def client_ip(request):
    cf = request.headers.get("CF-Connecting-IP")
    if cf:
        return cf
    xff = request.headers.get("X-Forwarded-For")
    if xff:
        return xff.split(",")[0].strip()
    return request.META.get("REMOTE_ADDR", "")


def allow(request, *, scope, limit, window, now=time.monotonic):
    """Return True if this (scope, client-ip) is within `limit` requests per
    `window` seconds; False if it has hit the limit. Counts the request when True."""
    t = now()
    if len(_buckets) > _MAX_KEYS:  # opportunistic eviction, by each bucket's own window
        for k, b in list(_buckets.items()):
            if t - b[0] >= b[2]:
                del _buckets[k]
    key = (scope, client_ip(request))
    bucket = _buckets.get(key)
    if bucket is None or t - bucket[0] >= window:
        _buckets[key] = [t, 1, window]
        return True
    if bucket[1] >= limit:
        return False
    bucket[1] += 1
    return True
